oscarapi/views/basket.py

```diff
@@ -60,51 +60,60 @@ def add_product(request, format=None):
         }]
     }
     """
-    p_ser = serializers.AddProductSerializer(data=request.DATA,
-                                             context={'request': request})
-    if p_ser.is_valid():
-        basket = get_basket(request)
-
-        product = p_ser.object['url']
-        quantity = p_ser.object['quantity']
-        options = p_ser.object.get('options', [])
-
-        availability = basket.strategy.fetch_for_product(product).availability
+    def validate(self, basket, product, quantity):
+        availability = basket.strategy.fetch_for_product(
+            product).availability
 
         # check if product is available at all
         if not availability.is_available_to_buy:
-            return Response(
-                {'reason': availability.message}, status=status.HTTP_406_NOT_ACCEPTABLE)
+            return False, availability.message
 
         # check if we can buy this quantity
         allowed, message = availability.is_purchase_permitted(quantity)
         if not allowed:
-            return Response({'reason': message}, status=status.HTTP_406_NOT_ACCEPTABLE)
+            return False, message
 
         # check if there is a limit on amount
         allowed, message = basket.is_quantity_allowed(quantity)
         if not allowed:
-            return Response({'reason': message}, status=status.HTTP_406_NOT_ACCEPTABLE)
-
-        basket.add_product(product, quantity=quantity, options=options)
-        apply_offers(request, basket)
-        ser = serializers.BasketSerializer(
-            basket,  context={'request': request})
-        return Response(ser.data)
-
-    return Response({'reason': p_ser.errors}, status=status.HTTP_406_NOT_ACCEPTABLE)
+            return False, quantity
+        return True, None
+
+    def post(self, request, format=None):
+        p_ser = serializers.AddProductSerializer(
+            data=request.DATA, context={'request': request})
+        if p_ser.is_valid():
+            basket = get_basket(request)
+            product = p_ser.object['url']
+            quantity = p_ser.init_data.get('quantity')
```

Comment:
not ok, the serializer has been updated, so all properties are available on the cleaned and validated object:
```
product = p_ser.object['url']
quantity = p_ser.object['quantity'] <-- that one is wrong
options = p_ser.object.get('options', [])